src/Octoliner.h

```diff
@@ -1,102 +1,49 @@
  *
- * Defines: octoliner object - 8-point line-folowing sensor driver
- * � Amperka LLC (https://amperka.com, dev@amperka.com)
+ * Product page: https://amperka.ru/product/zelo-folow-line-sensor
+ * © Amperka LLC (https://amperka.com, dev@amperka.com)
  * 
  * Author: Vasily Basalaev <vasily@amperka.ru>
  * Refactored by: Yury Botov <by@amperka.com>
  * License: GPLv3, all text here must be included in any redistribution.
  */
 
+#include "GpioExpander.h"
 #include <Arduino.h>
+#include <Wire.h>
 
 #ifndef __OCTOLINER_H__
 #define __OCTOLINER_H__
 
-#ifndef INPUT_PULLDOWN
-#define INPUT_PULLDOWN 0x3
-#endif
-//INPUT_PULLDOWN
-
-#ifndef DEFAULT_GPIOEXP_ADDR
-#define DEFAULT_GPIOEXP_ADDR 42
-#endif
-// DEFAULT_GPIOEXP_ADDR
+class Octoliner : private GpioExpander {
+public:
+    Octoliner(uint8_t i2cAddress = 42);
+    void begin();
+    void begin(TwoWire* wire);
+    void setSensitivity(uint8_t sense);
+    int16_t analogRead(uint8_t sensor);
+    void analogReadAll(int16_t* analogValues);
+    uint8_t digitalReadAll(void);
 
-// Pins definiion
-constexpr uint8_t LED_BRIGHTNESS_PIN = 9;
-constexpr uint8_t SENSITIVITY_PIN = 0;
+    float trackLine(void);
+    float trackLine(uint8_t pattern);
+    float trackLine(int16_t* analogValues);
 
-enum IOcommand {
-    WHO_AM_I //get UID
-    ,
-    RESET // reset
-    ,
-    CHANGE_I2C_ADDR // change I2C-address manually
-    ,
-    SAVE_I2C_ADDR // Save current address to flash
-    ,
-    PORT_MODE_INPUT // set pins to INPUT mode
-    ,
-    PORT_MODE_PULLUP // ... with pull-up internal resistor
-    ,
-    PORT_MODE_PULLDOWN // .. with pull-down internal resistor
-    ,
-    PORT_MODE_OUTPUT // set pins to OUTPUT mode
-    ,
-    DIGITAL_READ // read port (all of 8 pins)
-    ,
-    DIGITAL_WRITE_HIGH //
-    ,
-    DIGITAL_WRITE_LOW //
-    ,
-    ANALOG_WRITE //
-    ,
-    ANALOG_READ //
-    ,
-    PWM_FREQ // common pwm frequency to all pins
-    ,
-    ADC_SPEED,
-    SEND_MASTER_READED_UID,
-    CHANGE_I2C_ADDR_IF_UID_OK
-};
+    virtual uint8_t mapAnalogToPattern(int16_t* analogValues) const;
+    virtual float mapPatternToLine(uint8_t pattern) const;
 
-class Octoliner {
-public:
-    Octoliner();
-    Octoliner(uint8_t i2caddress);
-    void begin(uint8_t value);
-    void begin(void);
-    void digitalWrite(int pin, bool value);
-    void pinMode(int pin, uint8_t mode);
-    void analogWrite(int pin, uint8_t pulseWidth);
-    void pwmFreq(uint16_t freq);
     void changeAddr(uint8_t newAddr);
-    void changeAddrWithUID(uint8_t newAddr);
-    void saveAddr();
-    void reset();
-    int digitalRead(int pin);
-    int analogRead(int pin);
-    uint32_t getUID();
-    int digitalReadPort();
-    void digitalWritePort(uint16_t value);
-    void pinModePort(uint16_t value, uint8_t mode);
-    void adcSpeed(uint8_t speed);
-    void setSensitivity(uint8_t sense);
-    void setBrightness(uint8_t brightness);
-    float mapLine(int binaryLine[8]);
+    void saveAddr(void);
```

Comment:
Ouch… These are `changeAddress` and `saveAddress` in `API.md`. Suggest renaming to match the docs.